mlr-org/gallery/optimization/2022-11-30-hyperband-subsampling/index.qmd

```diff
@@ -0,0 +1,177 @@
+We connect the `r ref("PipeOp")`  with the learner and get a `r ref("GraphLearner")`.
+
+```{r}
+graph_learner = as_learner(
+  po("subsample") %>>%
+  learner
+)
+```
+
+The graph learner subsamples and then fits a support vector machine on the data subset.
+The parameter set of the graph learner is a combination of the parameter set of the `r ref("PipeOp")` and learner.
+
+```{r}
+as.data.table(graph_learner$param_set)[, .(id, lower, upper, levels)]
+```
+
+Next, we create the search space.
+We use `r ref("TuneToken")` to indicate that a hyperparameter should be tuned.
+We have to prefix the hyperparameters with the id of the `r ref("PipeOp", "PipeOps")`.
+The `subsample.frac` is the fidelity parameter that must be tagged with `"budget"` in the search space.
+The data set size is increased from 12.5% to 100%.
+For the other hyperparameters, we took the search space for support vector machines from the @kuehn_2018 article.
+This search space works for a wide range of data sets.
+
+```{r}
+graph_learner$param_set$set_values(
+  subsample.frac  = to_tune(p_dbl(2^-3, 1, tags = "budget")),
+  svm.kernel      = to_tune(c("linear", "polynomial", "radial")),
+  svm.cost        = to_tune(1e-4, 1e3, logscale = TRUE),
+  svm.gamma       = to_tune(1e-4, 1e3, logscale = TRUE),
+  svm.tolerance   = to_tune(1e-4, 2, logscale = TRUE),
+  svm.degree      = to_tune(2, 5)
+)
+```
+
+Let's create the tuning instance.
+We use the `"none"` terminator because Hyperband controls the termination itself.
+
+```{r}
+instance = ti(
+  task = tsk("sonar"),
+  learner = graph_learner,
+  resampling = rsmp("holdout"),
+  measures = msr("classif.ce"),
+  terminator = trm("none")
+)
+instance
+```
+
+We load the Hyperband `r ref("Tuner")` and set `eta = 2`.
+
+```{r}
+#| message: false
+library("mlr3hyperband")
+
+tuner = tnr("hyperband", eta = 2)
+```
+
+Using `eta = 2` and a lower bound of 12.5% for the data set size, results in the following schedule.
+Configurations with the same data set size are evaluated in parallel.
+
+```{r}
+#| echo: false
+library(htmltools)
+library(reactable)
+
+data = hyperband_schedule(r_min = 2^-3, r_max = 1, eta = 2)
+setorder(data, cols = "budget")
+
+
+reactable(data[, .(budget, bracket, stage, n)],
+  columns = list(
+    budget = colDef(
+      name = "Data Set Size",
+      cell = function(value) {
+        width = paste0(value * 100, "%")
+        value = format(width, width = 9, justify = "right")
+        bar = div(
+          class = "bar-chart",
+          style = list(marginRight = "0.375rem"),
+          div(class = "bar-chart__bar", style = list(width = width))
+        )
+        div(class = "bar-chart__bar-cell", span(class = "bar-chart__number", value), bar)
+      }
+    ),
+    bracket = colDef(name = "Bracket"),
+    stage = colDef(name = "Stage"),
+    n = colDef(name = "# Configruations")
+  ),
+  pagination = FALSE
+  )
+```
+
+Now we are ready to start the tuning.
+
+```{r}
+tuner$optimize(instance)
+```
+
```

Comment:
Kann man hier irgendwo noch eine Grafik zeigen die die Perf. in Abhängigkeit vom fidelity parameter zeigt?